File: python/map_set_bitmap.py
from typing import TypeVar, Generic, Optional, List, Tuple, Callable, Iterable, Any

K = TypeVar('K')
V = TypeVar('V')
T = TypeVar('T')


class HashNode(Generic[K, V]):
    def __init__(self, key: K, value: V):
        self.key: K = key
        self.value: V = value
        self.next: Optional[HashNode[K, V]] = None


class Map(Generic[K, V]):
    def __init__(self, initial_capacity: int = 16):
        self._capacity: int = initial_capacity
        self._size: int = 0
        self._buckets: List[Optional[HashNode[K, V]]] = [None] * initial_capacity
        self._load_factor: float = 0.7

    @property
    def size(self) -> int:
        return self._size

    @property
    def is_empty(self) -> bool:
        return self._size == 0

    @property
    def capacity(self) -> int:
        return self._capacity

    def _hash(self, key: K) -> int:
        hash_val = 0
        key_str = str(key)
        for char in key_str:
            hash_val = (hash_val << 5) - hash_val + ord(char)
            hash_val = hash_val & hash_val
        return abs(hash_val) % self._capacity

    def _resize(self) -> None:
        old_buckets = self._buckets
        self._capacity *= 2
        self._size = 0
        self._buckets = [None] * self._capacity

        for bucket in old_buckets:
            current = bucket
            while current:
                self.set(current.key, current.value)
                current = current.next

    def set(self, key: K, value: V) -> None:
        if self._size / self._capacity >= self._load_factor:
            self._resize()

        index = self._hash(key)
        current = self._buckets[index]

        while current:
            if current.key == key:
                current.value = value
                return
            current = current.next

        new_node = HashNode(key, value)
        new_node.next = self._buckets[index]
        self._buckets[index] = new_node
        self._size += 1

    def get(self, key: K) -> Optional[V]:
        index = self._hash(key)
        current = self._buckets[index]

        while current:
            if current.key == key:
                return current.value
            current = current.next

        return None

    def has(self, key: K) -> bool:
        return self.get(key) is not None

    def delete(self, key: K) -> bool:
        index = self._hash(key)
        current = self._buckets[index]
        prev: Optional[HashNode[K, V]] = None

        while current:
            if current.key == key:
                if prev:
                    prev.next = current.next
                else:
                    self._buckets[index] = current.next
                self._size -= 1
                return True
            prev = current
            current = current.next

        return False

    def keys(self) -> List[K]:
        keys_list: List[K] = []
        for bucket in self._buckets:
            current = bucket
            while current:
                keys_list.append(current.key)
                current = current.next
        return keys_list

    def values(self) -> List[V]:
        values_list: List[V] = []
        for bucket in self._buckets:
            current = bucket
            while current:
                values_list.append(current.value)
                current = current.next
        return values_list

    def entries(self) -> List[Tuple[K, V]]:
        entries_list: List[Tuple[K, V]] = []
        for bucket in self._buckets:
            current = bucket
            while current:
                entries_list.append((current.key, current.value))
                current = current.next
        return entries_list

    def clear(self) -> None:
        self._buckets = [None] * self._capacity
        self._size = 0

    def for_each(self, callback: Callable[[V, K, 'Map[K, V]'], None]) -> None:
        for bucket in self._buckets:
            current = bucket
            while current:
                callback(current.value, current.key, self)
                current = current.next

    def __getitem__(self, key: K) -> Optional[V]:
        return self.get(key)

    def __setitem__(self, key: K, value: V) -> None:
        self.set(key, value)

    def __contains__(self, key: K) -> bool:
        return self.has(key)

    def __len__(self) -> int:
        return self._size

    def __iter__(self):
        for key in self.keys():
            yield key

    def __repr__(self) -> str:
        items = [f"{repr(k)}: {repr(v)}" for k, v in self.entries()]
        return f"Map({{{', '.join(items)}}})"

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Map):
            return False
        if self.size != other.size:
            return False
        for key in self:
            if key not in other or self[key] != other[key]:
                return False
        return True

    def update(self, other: 'Map[K, V]') -> None:
        for key, value in other.entries():
            self[key] = value

    def __add__(self, other: 'Map[K, V]') -> 'Map[K, V]':
        result = Map[K, V](max(self.capacity, other.capacity))
        result.update(self)
        result.update(other)
        return result


class BitMap:
    def __init__(self, size: int = 64):
        self._size = size
        self._bits = [0] * ((size + 63) // 64)

    @property
    def size(self) -> int:
        return self._size

    def _get_index_and_mask(self, bit: int) -> Tuple[int, int]:
        if bit < 0 or bit >= self._size:
            raise IndexError(f"Bit index out of range: {bit}")
        index = bit // 64
        mask = 1 << (bit % 64)
        return index, mask

    def set(self, bit: int) -> None:
        index, mask = self._get_index_and_mask(bit)
        self._bits[index] |= mask

    def clear(self, bit: int) -> None:
        index, mask = self._get_index_and_mask(bit)
        self._bits[index] &= ~mask

    def toggle(self, bit: int) -> None:
        index, mask = self._get_index_and_mask(bit)
        self._bits[index] ^= mask

    def get(self, bit: int) -> bool:
        index, mask = self._get_index_and_mask(bit)
        return (self._bits[index] & mask) != 0

    def set_all(self) -> None:
        for i in range(len(self._bits)):
            self._bits[i] = 0xFFFFFFFFFFFFFFFF

    def clear_all(self) -> None:
        for i in range(len(self._bits)):
            self._bits[i] = 0

    def count_set_bits(self) -> int:
        count = 0
        for word in self._bits:
            count += bin(word).count('1')
        return count

    def find_first_set(self) -> Optional[int]:
        for i in range(self._size):
            if self.get(i):
                return i
        return None

    def find_first_clear(self) -> Optional[int]:
        for i in range(self._size):
            if not self.get(i):
                return i
        return None

    def __getitem__(self, bit: int) -> bool:
        return self.get(bit)

    def __setitem__(self, bit: int, value: bool) -> None:
        if value:
            self.set(bit)
        else:
            self.clear(bit)

    def __and__(self, other: 'BitMap') -> 'BitMap':
        if self._size != other._size:
            raise ValueError("BitMaps must have the same size for AND operation")
        result = BitMap(self._size)
        for i in range(len(self._bits)):
            result._bits[i] = self._bits[i] & other._bits[i]
        return result

    def __or__(self, other: 'BitMap') -> 'BitMap':
        if self._size != other._size:
            raise ValueError("BitMaps must have the same size for OR operation")
        result = BitMap(self._size)
        for i in range(len(self._bits)):
            result._bits[i] = self._bits[i] | other._bits[i]
        return result

    def __xor__(self, other: 'BitMap') -> 'BitMap':
        if self._size != other._size:
            raise ValueError("BitMaps must have the same size for XOR operation")
        result = BitMap(self._size)
        for i in range(len(self._bits)):
            result._bits[i] = self._bits[i] ^ other._bits[i]
        return result

    def __invert__(self) -> 'BitMap':
        result = BitMap(self._size)
        for i in range(len(self._bits)):
            result._bits[i] = ~self._bits[i]
        return result

    def __len__(self) -> int:
        return self._size

    def __iter__(self):
        for i in range(self._size):
            yield self.get(i)

    def __repr__(self) -> str:
        bits_str = ''.join('1' if self.get(i) else '0' for i in range(self._size))
        return f"BitMap({bits_str})"


class Set(Generic[T]):
    def __init__(self, initial_capacity: int = 16):
        self._map: Map[T, bool] = Map(initial_capacity)

    @property
    def size(self) -> int:
        return self._map.size

    @property
    def is_empty(self) -> bool:
        return self._map.is_empty

    def add(self, item: T) -> None:
        self._map.set(item, True)

    def remove(self, item: T) -> bool:
        return self._map.delete(item)

    def has(self, item: T) -> bool:
        return self._map.has(item)

    def clear(self) -> None:
        self._map.clear()

    def items(self) -> List[T]:
        return self._map.keys()

    def for_each(self, callback: Callable[[T, 'Set[T]'], None]) -> None:
        def map_callback(value: bool, key: T, map_obj: Map[T, bool]) -> None:
            callback(key, self)
        self._map.for_each(map_callback)

    def union(self, other: 'Set[T]') -> 'Set[T]':
        result = Set[T](max(self.size, other.size) + 1)
        for item in self:
            result.add(item)
        for item in other:
            result.add(item)
        return result

    def intersection(self, other: 'Set[T]') -> 'Set[T]':
        result = Set[T]()
        smaller = self if self.size <= other.size else other
        larger = other if self.size <= other.size else self
        for item in smaller:
            if larger.has(item):
                result.add(item)
        return result

    def difference(self, other: 'Set[T]') -> 'Set[T]':
        result = Set[T]()
        for item in self:
            if not other.has(item):
                result.add(item)
        return result

    def symmetric_difference(self, other: 'Set[T]') -> 'Set[T]':
        result = Set[T]()
        for item in self:
            if not other.has(item):
                result.add(item)
        for item in other:
            if not self.has(item):
                result.add(item)
        return result

    def is_subset(self, other: 'Set[T]') -> bool:
        if self.size > other.size:
            return False
        for item in self:
            if not other.has(item):
                return False
        return True

    def is_superset(self, other: 'Set[T]') -> bool:
        return other.is_subset(self)

    def __contains__(self, item: T) -> bool:
        return self.has(item)

    def __len__(self) -> int:
        return self.size

    def __iter__(self):
        for item in self.items():
            yield item

    def __repr__(self) -> str:
        items = [repr(item) for item in self.items()]
        return f"Set({{{', '.join(items)}}})"

    def __add__(self, other: 'Set[T]') -> 'Set[T]':
        return self.union(other)

    def __sub__(self, other: 'Set[T]') -> 'Set[T]':
        return self.difference(other)

    def __and__(self, other: 'Set[T]') -> 'Set[T]':
        return self.intersection(other)

    def __or__(self, other: 'Set[T]') -> 'Set[T]':
        return self.union(other)

    def __xor__(self, other: 'Set[T]') -> 'Set[T]':
        return self.symmetric_difference(other)

    def __le__(self, other: 'Set[T]') -> bool:
        return self.is_subset(other)

    def __ge__(self, other: 'Set[T]') -> bool:
        return self.is_superset(other)

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Set):
            return False
        if self.size != other.size:
            return False
        return self.is_subset(other)


if __name__ == "__main__":
    print("=" * 50)
    print("Map 示例:")
    print("=" * 50)
    map1 = Map[str, int]()
    map1["one"] = 1
    map1["two"] = 2
    map1["three"] = 3
    print(f"Map: {map1}")
    print(f"大小: {len(map1)}")
    print(f"包含 'two': {'two' in map1}")
    print(f"'two' 的值: {map1['two']}")
    print(f"键: {map1.keys()}")
    print(f"值: {map1.values()}")
    print()

    print("=" * 50)
    print("BitMap 示例:")
    print("=" * 50)
    bitmap = BitMap(10)
    bitmap[0] = True
    bitmap[2] = True
    bitmap[5] = True
    print(f"BitMap: {bitmap}")
    print(f"设置的位数: {bitmap.count_set_bits()}")
    print(f"第 2 位: {bitmap[2]}")
    bitmap.toggle(2)
    print(f"翻转第 2 位后: {bitmap}")
    print()

    print("=" * 50)
    print("Set 示例:")
    print("=" * 50)
    set1 = Set[int]()
    set1.add(1)
    set1.add(2)
    set1.add(3)
    set2 = Set[int]()
    set2.add(3)
    set2.add(4)
    set2.add(5)
    print(f"Set1: {set1}")
    print(f"Set2: {set2}")
    print(f"并集: {set1 | set2}")
    print(f"交集: {set1 & set2}")
    print(f"差集 (Set1 - Set2): {set1 - set2}")
    print(f"包含 2: {2 in set1}")

File: go/map_set_bitmap.go
package main

import (
	"errors"
	"fmt"
	"math/bits"
)

type MapHashNode[K comparable, V any] struct {
	key   K
	value V
	next  *MapHashNode[K, V]
}

type Map[K comparable, V any] struct {
	capacity   int
	size       int
	buckets    []*MapHashNode[K, V]
	loadFactor float64
}

func NewMap[K comparable, V any](initialCapacity int) (*Map[K, V], error) {
	if initialCapacity <= 0 {
		return nil, errors.New("initial capacity must be positive")
	}
	return &Map[K, V]{
		capacity:   initialCapacity,
		size:       0,
		buckets:    make([]*MapHashNode[K, V], initialCapacity),
		loadFactor: 0.7,
	}, nil
}

func (m *Map[K, V]) Size() int {
	return m.size
}

func (m *Map[K, V]) IsEmpty() bool {
	return m.size == 0
}

func (m *Map[K, V]) Capacity() int {
	return m.capacity
}

func (m *Map[K, V]) hash(key K) int {
	keyStr := fmt.Sprintf("%v", key)
	hash := 0
	for _, char := range keyStr {
		hash = (hash << 5) - hash + int(char)
	}
	if hash < 0 {
		hash = -hash
	}
	return hash % m.capacity
}

func (m *Map[K, V]) resize() {
	oldBuckets := m.buckets
	m.capacity *= 2
	m.size = 0
	m.buckets = make([]*MapHashNode[K, V], m.capacity)

	for _, bucket := range oldBuckets {
		current := bucket
		for current != nil {
			m.Set(current.key, current.value)
			current = current.next
		}
	}
}

func (m *Map[K, V]) Set(key K, value V) {
	if float64(m.size)/float64(m.capacity) >= m.loadFactor {
		m.resize()
	}

	index := m.hash(key)
	current := m.buckets[index]

	for current != nil {
		if current.key == key {
			current.value = value
			return
		}
		current = current.next
	}

	newNode := &MapHashNode[K, V]{key: key, value: value, next: m.buckets[index]}
	m.buckets[index] = newNode
	m.size++
}

func (m *Map[K, V]) Get(key K) (V, error) {
	var zero V
	index := m.hash(key)
	current := m.buckets[index]

	for current != nil {
		if current.key == key {
			return current.value, nil
		}
		current = current.next
	}

	return zero, errors.New("key not found")
}

func (m *Map[K, V]) Has(key K) bool {
	index := m.hash(key)
	current := m.buckets[index]

	for current != nil {
		if current.key == key {
			return true
		}
		current = current.next
	}

	return false
}

func (m *Map[K, V]) Delete(key K) bool {
	index := m.hash(key)
	current := m.buckets[index]
	var prev *MapHashNode[K, V] = nil

	for current != nil {
		if current.key == key {
			if prev != nil {
				prev.next = current.next
			} else {
				m.buckets[index] = current.next
			}
			m.size--
			return true
		}
		prev = current
		current = current.next
	}

	return false
}

func (m *Map[K, V]) Keys() []K {
	keys := make([]K, 0, m.size)
	for _, bucket := range m.buckets {
		current := bucket
		for current != nil {
			keys = append(keys, current.key)
			current = current.next
		}
	}
	return keys
}

func (m *Map[K, V]) Values() []V {
	values := make([]V, 0, m.size)
	for _, bucket := range m.buckets {
		current := bucket
		for current != nil {
			values = append(values, current.value)
			current = current.next
		}
	}
	return values
}

func (m *Map[K, V]) Entries() [][2]interface{} {
	entries := make([][2]interface{}, 0, m.size)
	for _, bucket := range m.buckets {
		current := bucket
		for current != nil {
			entries = append(entries, [2]interface{}{current.key, current.value})
			current = current.next
		}
	}
	return entries
}

func (m *Map[K, V]) Clear() {
	m.buckets = make([]*MapHashNode[K, V], m.capacity)
	m.size = 0
}

func (m *Map[K, V]) Update(other *Map[K, V]) {
	for _, entry := range other.Entries() {
		key, _ := entry[0].(K)
		value, _ := entry[1].(V)
		m.Set(key, value)
	}
}

// MapIterator 是Map的迭代器
type MapIterator[K comparable, V any] struct {
	mapObj      *Map[K, V]
	bucketIndex int
	current     *MapHashNode[K, V]
}

// Iterator 返回Map的迭代器
func (m *Map[K, V]) Iterator() *MapIterator[K, V] {
	return &MapIterator[K, V]{
		mapObj:      m,
		bucketIndex: -1,
		current:     nil,
	}
}

// Next 移动到下一个元素
func (it *MapIterator[K, V]) Next() bool {
	if it.current != nil {
		it.current = it.current.next
		if it.current != nil {
			return true
		}
	}

	it.bucketIndex++
	for it.bucketIndex < it.mapObj.capacity {
		if it.mapObj.buckets[it.bucketIndex] != nil {
			it.current = it.mapObj.buckets[it.bucketIndex]
			return true
		}
		it.bucketIndex++
	}

	return false
}

// Key 返回当前元素的键
func (it *MapIterator[K, V]) Key() K {
	return it.current.key
}

// Value 返回当前元素的值
func (it *MapIterator[K, V]) Value() V {
	return it.current.value
}

// Entry 返回当前元素的键值对
func (it *MapIterator[K, V]) Entry() (K, V) {
	return it.current.key, it.current.value
}

// ForEach 遍历Map中的所有元素
func (m *Map[K, V]) ForEach(f func(K, V)) {
	for _, bucket := range m.buckets {
		current := bucket
		for current != nil {
			f(current.key, current.value)
			current = current.next
		}
	}
}

type BitMap struct {
	size int
	bits []uint64
}

func NewBitMap(size int) (*BitMap, error) {
	if size <= 0 {
		return nil, errors.New("size must be positive")
	}
	numWords := (size + 63) / 64
	return &BitMap{
		size: size,
		bits: make([]uint64, numWords),
	}, nil
}

func (bm *BitMap) Size() int {
	return bm.size
}

func (bm *BitMap) getIndexAndMask(bit int) (int, uint64, error) {
	if bit < 0 || bit >= bm.size {
		return 0, 0, errors.New("bit index out of bounds")
	}
	index := bit / 64
	mask := uint64(1) << (bit % 64)
	return index, mask, nil
}

func (bm *BitMap) Set(bit int) error {
	index, mask, err := bm.getIndexAndMask(bit)
	if err != nil {
		return err
	}
	bm.bits[index] |= mask
	return nil
}

func (bm *BitMap) Clear(bit int) error {
	index, mask, err := bm.getIndexAndMask(bit)
	if err != nil {
		return err
	}
	bm.bits[index] &^= mask
	return nil
}

func (bm *BitMap) Toggle(bit int) error {
	index, mask, err := bm.getIndexAndMask(bit)
	if err != nil {
		return err
	}
	bm.bits[index] ^= mask
	return nil
}

func (bm *BitMap) Get(bit int) (bool, error) {
	index, mask, err := bm.getIndexAndMask(bit)
	if err != nil {
		return false, err
	}
	return (bm.bits[index] & mask) != 0, nil
}

func (bm *BitMap) SetAll() {
	for i := range bm.bits {
		bm.bits[i] = ^uint64(0)
	}
}

func (bm *BitMap) ClearAll() {
	for i := range bm.bits {
		bm.bits[i] = 0
	}
}

func (bm *BitMap) CountSetBits() int {
	count := 0
	for _, word := range bm.bits {
		count += bits.OnesCount64(word)
	}
	return count
}

func (bm *BitMap) FindFirstSet() (int, error) {
	for i := 0; i < bm.size; i++ {
		set, err := bm.Get(i)
		if err != nil {
			return -1, err
		}
		if set {
			return i, nil
		}
	}
	return -1, nil
}

func (bm *BitMap) FindFirstClear() (int, error) {
	for i := 0; i < bm.size; i++ {
		set, err := bm.Get(i)
		if err != nil {
			return -1, err
		}
		if !set {
			return i, nil
		}
	}
	return -1, nil
}

func (bm *BitMap) And(other *BitMap) (*BitMap, error) {
	if bm.size != other.size {
		return nil, errors.New("BitMaps must have the same size")
	}
	result, _ := NewBitMap(bm.size)
	for i := range bm.bits {
		result.bits[i] = bm.bits[i] & other.bits[i]
	}
	return result, nil
}

func (bm *BitMap) Or(other *BitMap) (*BitMap, error) {
	if bm.size != other.size {
		return nil, errors.New("BitMaps must have the same size")
	}
	result, _ := NewBitMap(bm.size)
	for i := range bm.bits {
		result.bits[i] = bm.bits[i] | other.bits[i]
	}
	return result, nil
}

func (bm *BitMap) Xor(other *BitMap) (*BitMap, error) {
	if bm.size != other.size {
		return nil, errors.New("BitMaps must have the same size")
	}
	result, _ := NewBitMap(bm.size)
	for i := range bm.bits {
		result.bits[i] = bm.bits[i] ^ other.bits[i]
	}
	return result, nil
}

func (bm *BitMap) Not() (*BitMap, error) {
	result, _ := NewBitMap(bm.size)
	for i := range bm.bits {
		result.bits[i] = ^bm.bits[i]
	}
	return result, nil
}

func (bm *BitMap) String() string {
	bitsStr := make([]byte, bm.size)
	for i := 0; i < bm.size; i++ {
		set, _ := bm.Get(i)
		if set {
			bitsStr[i] = '1'
		} else {
			bitsStr[i] = '0'
		}
	}
	return string(bitsStr)
}

// BitMapIterator 是BitMap的迭代器
type BitMapIterator struct {
	bitmap  *BitMap
	current int
}

// Iterator 返回BitMap的迭代器
func (bm *BitMap) Iterator() *BitMapIterator {
	return &BitMapIterator{
		bitmap:  bm,
		current: -1,
	}
}

// Next 移动到下一个元素
func (it *BitMapIterator) Next() bool {
	it.current++
	return it.current < it.bitmap.size
}

// Index 返回当前位的索引
func (it *BitMapIterator) Index() int {
	return it.current
}

// Value 返回当前位的值
func (it *BitMapIterator) Value() (bool, error) {
	return it.bitmap.Get(it.current)
}

// ForEach 遍历BitMap中的所有位
func (bm *BitMap) ForEach(f func(int, bool)) {
	for i := 0; i < bm.size; i++ {
		set, _ := bm.Get(i)
		f(i, set)
	}
}

type Set[T comparable] struct {
	m *Map[T, bool]
}

func NewSet[T comparable](initialCapacity int) (*Set[T], error) {
	m, err := NewMap[T, bool](initialCapacity)
	if err != nil {
		return nil, err
	}
	return &Set[T]{m: m}, nil
}

func (s *Set[T]) Size() int {
	return s.m.Size()
}

func (s *Set[T]) IsEmpty() bool {
	return s.m.IsEmpty()
}

func (s *Set[T]) Add(item T) {
	s.m.Set(item, true)
}

func (s *Set[T]) Remove(item T) bool {
	return s.m.Delete(item)
}

func (s *Set[T]) Has(item T) bool {
	return s.m.Has(item)
}

func (s *Set[T]) Clear() {
	s.m.Clear()
}

func (s *Set[T]) Items() []T {
	return s.m.Keys()
}

func (s *Set[T]) Union(other *Set[T]) (*Set[T], error) {
	result, _ := NewSet[T](mapSetMax(s.Size(), other.Size()) + 1)
	for _, item := range s.Items() {
		result.Add(item)
	}
	for _, item := range other.Items() {
		result.Add(item)
	}
	return result, nil
}

func (s *Set[T]) Intersection(other *Set[T]) (*Set[T], error) {
	result, _ := NewSet[T](16)
	smaller, larger := s, other
	if s.Size() > other.Size() {
		smaller, larger = other, s
	}
	for _, item := range smaller.Items() {
		if larger.Has(item) {
			result.Add(item)
		}
	}
	return result, nil
}

func (s *Set[T]) Difference(other *Set[T]) (*Set[T], error) {
	result, _ := NewSet[T](16)
	for _, item := range s.Items() {
		if !other.Has(item) {
			result.Add(item)
		}
	}
	return result, nil
}

func (s *Set[T]) SymmetricDifference(other *Set[T]) (*Set[T], error) {
	result, _ := NewSet[T](16)
	for _, item := range s.Items() {
		if !other.Has(item) {
			result.Add(item)
		}
	}
	for _, item := range other.Items() {
		if !s.Has(item) {
			result.Add(item)
		}
	}
	return result, nil
}

func (s *Set[T]) IsSubset(other *Set[T]) bool {
	if s.Size() > other.Size() {
		return false
	}
	for _, item := range s.Items() {
		if !other.Has(item) {
			return false
		}
	}
	return true
}

func (s *Set[T]) IsSuperset(other *Set[T]) bool {
	return other.IsSubset(s)
}

func (s *Set[T]) String() string {
	items := make([]string, 0, s.Size())
	for _, item := range s.Items() {
		items = append(items, fmt.Sprintf("%v", item))
	}
	return fmt.Sprintf("{%s}", mapSetJoin(items, ", "))
}

// SetIterator 是Set的迭代器
type SetIterator[T comparable] struct {
	set      *Set[T]
	iterator *MapIterator[T, bool]
}

// Iterator 返回Set的迭代器
func (s *Set[T]) Iterator() *SetIterator[T] {
	return &SetIterator[T]{
		set:      s,
		iterator: s.m.Iterator(),
	}
}

// Next 移动到下一个元素
func (it *SetIterator[T]) Next() bool {
	return it.iterator.Next()
}

// Value 返回当前元素的值
func (it *SetIterator[T]) Value() T {
	return it.iterator.Key()
}

// ForEach 遍历Set中的所有元素
func (s *Set[T]) ForEach(f func(T)) {
	s.m.ForEach(func(key T, _ bool) {
		f(key)
	})
}

func mapSetMax(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func mapSetJoin(items []string, sep string) string {
	if len(items) == 0 {
		return ""
	}
	result := items[0]
	for i := 1; i < len(items); i++ {
		result += sep + items[i]
	}
	return result
}

// func main() {
// 	// 测试Map的可遍历结构
// 	fmt.Println("=== 测试Map的可遍历结构 ===")
// 	m, _ := NewMap[string, int](16)
// 	m.Set("one", 1)
// 	m.Set("two", 2)
// 	m.Set("three", 3)

// 	// 使用ForEach方法遍历
// 	fmt.Println("使用ForEach方法遍历Map:")
// 	m.ForEach(func(key string, value int) {
// 		fmt.Printf("%s: %d\n", key, value)
// 	})

// 	// 使用迭代器遍历
// 	fmt.Println("使用迭代器遍历Map:")
// 	it := m.Iterator()
// 	for it.Next() {
// 		fmt.Printf("%s: %d\n", it.Key(), it.Value())
// 	}

// 	// 测试Set的可遍历结构
// 	fmt.Println("\n=== 测试Set的可遍历结构 ===")
// 	s, _ := NewSet[int](16)
// 	s.Add(1)
// 	s.Add(2)
// 	s.Add(3)

// 	// 使用ForEach方法遍历
// 	fmt.Println("使用ForEach方法遍历Set:")
// 	s.ForEach(func(item int) {
// 		fmt.Printf("%d\n", item)
// 	})

// 	// 使用迭代器遍历
// 	fmt.Println("使用迭代器遍历Set:")
// 	sit := s.Iterator()
// 	for sit.Next() {
// 		fmt.Printf("%d\n", sit.Value())
// 	}

// 	// 测试BitMap的可遍历结构
// 	fmt.Println("\n=== 测试BitMap的可遍历结构 ===")
// 	bm, _ := NewBitMap(10)
// 	bm.Set(0)
// 	bm.Set(2)
// 	bm.Set(5)

// 	// 使用ForEach方法遍历
// 	fmt.Println("使用ForEach方法遍历BitMap:")
// 	bm.ForEach(func(index int, value bool) {
// 		fmt.Printf("%d: %t\n", index, value)
// 	})

// 	// 使用迭代器遍历
// 	fmt.Println("使用迭代器遍历BitMap:")
// 	bmit := bm.Iterator()
// 	for bmit.Next() {
// 		value, _ := bmit.Value()
// 		fmt.Printf("%d: %t\n", bmit.Index(), value)
// 	}
// }
